Adds documentation around the halibut protocol. (#205)
OctopusDeploy/Halibut: docs/halibut-protocol.md

> # Halibut Protocol data exchange

Author: Updated to make mention of the control messages during initiation as well as some hints at what the messages are trying to signal.

> # Halibut Protocol data exchange

Reviewer: Starting to look good, although I'm finding the numbered steps with the "Steps 2 to 5 are repeated..." slightly hard to follow. 

I wonder if it might be helpful to split the numbered list into 3 sub-sections to help show the various phases of the connection:
- Identification
- Message exchange
- Ending a connection

Having separate sub-sections might allow more room to expand on some things like the difference between graceful and non-graceful termination of the connection (e.g. do both parties send the `END` message?)

What do you think?

> # Halibut Protocol data exchange

Author: I tried to do this, did it work?

> # Halibut Protocol data exchange

Reviewer: Yep that reads much better, thanks.

When invoking a remote method in both polling and listening mode, the same general process is followed:

>  - The Server sends a `Request` message containing the method to execute and data.

Author: I think defining them upfront makes sense.

 - The client executes the method and sends the result in a `Response` message.
 - The client sends a `NEXT` control message.
 - The Server semds a `Proceed` control message.
 - The above 4 steps are repeated until the connection is terminated or control `END` messages are sent.

## Listening client protocol data exchange

![Listening client protocol data exchange](images/listeningprotocoldata.png)

## Polling client protocol data exchange

![Polling client protocol data exchange](images/pollingprotocoldata.png)

## Request and Response message format

Request and Response messages are sent and received in the same format with only the data within the BSON section changing to denote the type of message.

> 

Author: How is that?

> 

Reviewer: Mostly looks good, I've made a couple of suggestions inline for you to consider.

> ![Request/Response message format](images/message-format.png)

Reviewer: These images don't appear to be rendering all that well when I look at the [branch directly in GitHub](https://github.com/OctopusDeploy/Halibut/blob/luke/illustrate-the-halibut-protocol/docs/halibut-protocol.md), I wonder if they could do with a white background?

Author: I see them with a white background...

Author: Although for me they are blurry:

![image](https://user-images.githubusercontent.com/7076477/218373129-c60c87fa-5e2c-4d10-8dc6-a4b0135e5f9d.png)

Reviewer: I think they have a transparent background, I'm using a dark theme in GitHub so the background appears dark and some of the words are hard to read.

<img width="889" alt="image" src="https://user-images.githubusercontent.com/2915931/218374990-a8b2b43a-2a8d-4018-8b8e-eddd23f31a94.png">

Author: oh boy

Reviewer: If you re-export the pictures with a white background hopefully it should appear ok for both light and dark themes?

Author: try now?

Reviewer: Yep looking good now, thanks

# Halibut Protocol data exchange

This document outlines the data that is exchanged between the **Client** and **Service**. In this document:

 - **Client** is what makes the remote procedure calls e.g. Octopus.

>  - **Service** is what execiutes the methods e.g. Tentacle. 

Author: These terms are lifted from the Halibut readme.

When invoking a remote method in both polling and listening mode, the same general process is followed:

> 1.  After the TCP connection is intiated, both server and client identify themselves to each other through a set of control messages.

Reviewer: It might be worth mentioning that the identification is different for listening vs polling tentacles and adding a link to those sections.

Author: Done

2. The Client sends a `Request` message containing the method to execute and data.
3. The Service executes the method and sends the result in a `Response` message.
4. The Service sends a `NEXT` control message, signaling it is ready for the next action.

> 5. The Client sends a `Proceed` control message, signaling to the Service to be ready for another `Request`.

Reviewer: The casing is different here compared to `NEXT` and `END`, is that how Halibut works?

Author: nice catch

6. Steps 2 to 5 are repeated until the connection is terminated or control `END` messages are sent.

## Listening Service protocol data exchange

When the **Service** is in  **Listening** mode, the **Client** it identifies itself as `MX_CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.

![Listening client protocol data exchange](images/listeningprotocoldata.png)

## Polling Service protocol data exchange

When the **Service** is in  **Polling** mode, the **Service** identifies itself as `MX-SUBSCRIBER 1.0` while the `Client` identifies itself as `MX-SERVER 1.0`.

![Polling client protocol data exchange](images/pollingprotocoldata.png)

## Request and Response message format

> The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by multiple [DataStream](../source/Halibut/DataStream.cs). DataStreams is the way Halibut sends large amounts of data. DataStreams are not compressed and are send one by one after the compressed BSON. DataStreams are referenced by GUID in the BSON section and labelled with a GUID in the DataStream section.

Reviewer: ```suggestion
The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by zero or more [DataStream]s(../source/Halibut/DataStream.cs). A `DataStream` represents data that should not be serialized as part of a message, for example a file to be transferred. They can be sent in either a request or a response. `DataStream`s are transferred as raw bytes in the TCP stream (i.e. they are not compressed) and are sent sequentially after the compressed BSON of the request/response. Each `DataStream` has a unique GUID which is referenced in the request/response so that it can be used by calling code.
```

# Halibut Protocol data exchange

This document outlines the data that is exchanged between the **Client** and **Service**. In this document:

 - **Client** is what makes the remote procedure calls e.g. Octopus Server.
 - **Service** is what executes the methods e.g. Tentacle. 

When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.

 ### Stage 1 - Identification
 

>   A TCP connection is established and both server and client identify themselves, though a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.

Reviewer 2: ```suggestion
  A TCP connection is established and both server and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
```

### Stage 2 - Message exchange

Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:

1. The Client sends a `Request` message containing the method to execute and data.
2. The Service executes the method and sends the result in a `Response` message.
3. The Service sends a `NEXT` control message, signaling it is ready for the next action.
4. The Client sends a `PROCEED` control message, signaling to the Service to be ready for another `Request`.

> The steps are repeated for both Listening and Polling mode until either Client or Service begins the Ending a connection stage.

Reviewer 2: Is it worth mentioning that during polling, the `NEXT`/`PROCEED` cycle repeats itself periodically when no requests are being sent?

Author: Done under the polling section

### Stage 3 - Ending a connection

The final stage can be entered by either the Client or Service terminating the TCP connection or sending a `END` control message. Once this stage is reached no more requests will be processed in the current TCP connection.

## Listening Service protocol data exchange

> When the **Service** is in  **Listening** mode, the **Client** it identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.

Reviewer 2: ```suggestion
When the **Service** is in  **Listening** mode, the **Client** identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.
```

# Halibut Protocol data exchange

This document outlines the data that is exchanged between the **Client** and **Service**. In this document:

 - **Client** is what makes the remote procedure calls e.g. Octopus Server.
 - **Service** is what executes the methods e.g. Tentacle. 

When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.

 ### Stage 1 - Identification
 

>   A TCP connection is established and both server and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.

Reviewer 3: server -> service

Author: fixed

The steps are repeated for both Listening and Polling mode until either Client or Service begins the Ending a connection stage.

### Stage 3 - Ending a connection

The final stage can be entered by either the Client or Service terminating the TCP connection or sending a `END` control message. Once this stage is reached no more requests will be processed in the current TCP connection.

## Listening Service protocol data exchange

When the **Service** is in  **Listening** mode, the **Client** identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.

![Listening client protocol data exchange](images/listeningprotocoldata.png)

## Polling Service protocol data exchange

When the **Service** is in  **Polling** mode, the **Service** identifies itself as `MX-SUBSCRIBER 1.0` while the `Client` identifies itself as `MX-SERVER 1.0`.

Under polling mode, unlike listening, the Client will periodically send a no-op NUL Request to the Service to keep the TCP connection alive.

![Polling client protocol data exchange](images/pollingprotocoldata.png)

## Request and Response message format

> The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by zero or more [DataStream]s(../source/Halibut/DataStream.cs). A `DataStream` represents data that should not be serialized as part of a message, for example a file to be transferred. They can be sent in either a request or a response. `DataStream`s are transferred as raw bytes in the TCP stream (i.e. they are not compressed) and are sent sequentially after the compressed BSON of the request/response. Each `DataStream` has a unique GUID which is referenced in the request/response so that it can be used by calling code.

Reviewer 3: [DataStream]s(../source/Halibut/DataStream.cs) is not linking tot he file due to the rouge s

Author: Fixed

> # Halibut Protocol data exchange

Reviewer 4: Consider mentioning that the protocol is inspired by JSON-RPC and adding a link to https://www.codeproject.com/Articles/554692/Halibut-a-secure-communication-stack-for-NET-Mono to provide some historical context.

Author: added

This document outlines the data that is exchanged between the **Client** and **Service**. In this document:

 - **Client** is what makes the remote procedure calls e.g. Octopus Server.
 - **Service** is what executes the methods e.g. Tentacle. 

When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.

 ### Stage 1 - Identification
 
  A TCP connection is established and both Service and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.

### Stage 2 - Message exchange

Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:

> 1. The Client sends a `Request` message containing the method to execute and data.

Reviewer 4: `data` is a bit overload term in this context. What about `message containing the method to execute, its parameters and 0 or more data streams`

> 2. The Service executes the method and sends the result in a `Response` message.

Reviewer 4: Similar to above, "The Service executes the method and sends a `Response` message containing the result and 0 or more data streams`
